### ISSUE TITLE
I installed ExplorerPatcher but I couldn't activate certain options and now I can't fully uninstall it

### ISSUE THREAD

**Issue by LOuroboros** [valinet/ExplorerPatcher#3955]
Hello.

Earlier this morning, I did my yearly PC formatting and reinstalled Windows. I'm currently on Windows 11 24H2 26100.2314 coming from Windows 11 22H2 22621.2283. In that build everything was working perfectly fine and Explorer Patcher never gave me any major issues.

Now on my fresh install I did my usual process of installing drivers, redist C++ libraries and so on.

Every single thing worked out normally. Except that when I installed Explorer Patcher v22621.4317.67.1 I noticed that one of the options I always enable wasn't working anymore, the one that restored the squared corners for the different windows. Whenever I enabled it and restarted the File Explorer, the option would disable itself on its own and I could see that the corners of the windows were rounded.

I didn't think much of it. [I just used the standalone solution to get my squared corners back](https://github.com/valinet/Win11DisableRoundedCorners).

Then I noticed that whenever I changed the Window Switcher setting, none of the options besides "Windows 11 (Default)" were working.

Finally, I decided to just reinstall Explorer Patcher and hope that would get things working normally again, however...

![ep_setup_20241122_001923955](https://github.com/user-attachments/assets/752805fa-ea11-4b56-a479-a15d16c76ff8)

That is what I get after a bit of waiting whenever I try to uninstall Explorer Patcher.

Before uninstalling I used the "Restore Default Settings" option too, but I got an error telling me that the keys might have been used by an active process and such. I imagined a reboot would be good enough to deal with that, and it is after that reboot that I tried to uninstall Explorer Patcher.

Now I'm finding myself incapable of reinstalling the program because it doesn't uninstall itself completely, and I honestly don't know what to do.

After I got the error above I tried to uninstall yet again, but it seems to have gotten stuck in some invisible application or something?

![explorer_20241122_003044962](https://github.com/user-attachments/assets/e2626402-a33d-4595-9f28-e7af943ebbf3)

If there's any more information I can or should provide, please let me know.

**Comment by LOuroboros** [comment]
Brief status update; after writing this Issue ticket, I restarted the PC one more time and tried to uninstall Explorer Patcher yet again. 

Sadly, it didn't work.

![test1](https://github.com/user-attachments/assets/ea7d0cc8-8791-4ee3-bece-bcb4b009e3b0)

After I clicked "Yes", the Windows Explorer closed itself, and I was thrown this error which I already showed above.

![test2](https://github.com/user-attachments/assets/650198ba-3a15-4790-b070-f0e440bf9c2e)

So, my current situation is that the changes Explorer Patcher applied to my system have seemingly been reverted. Everything looks and works like stock Windows 11 *(except for my start menu, but that's because [I use Open Shell](https://github.com/Open-Shell/Open-Shell-Menu))*, but I can't fully uninstall Explorer Patcher from my system, and by extension I can't really reinstall it.

**Comment by LOuroboros** [comment]
> you probably didn't run the powershell commands mentioned here first: https://github.com/valinet/ExplorerPatcher/releases
> 
> you must run those before installing EP. Otherwise it will prevent parts of EP from being installed.
> 
> The powershell script you must run with elevated privileges (right click on powershell and select run as administrator).
> 
> Future updates to EP won't be flagged then.
> 
> You can also set windows defender to exclude the directory that you manually download EP to so you can install it without windows defender blocking it.
> 
> if you get the error: 0x800106ba, it means Windows Defender is not active. You may have another antivirus program active, which means you need to add those files and directories to be excluded in your antivirus program.
> 
> If that all succeeds, then you can install EP.
> 
> See here for how to uninstall EP after: https://github.com/valinet/ExplorerPatcher/wiki/Installer-How-To#how-to-uninstall
> 
> As for the rounded corners option not working, that is already known here: #3555

I do not have Windows Defender for reasons, so none of the files were actually flagged or anything, and I'm not using any other AV program either.

Right now, what I'd like to know is how to fully uninstall Explorer Patcher in my current situation where everything is working like a default installation of Windows 11, but Windows still refuses to let me fully uninstall the program.

**Comment by Amrsatrio** [comment]
Please go to C:\Program Files\ExplorerPatcher. Are there .prev files? Delete them if there are any. Then check %APPDATA%\ExplorerPatcher and check if there's a folder named cleanup. If there is one, what are the files inside? Try to delete them as well.

I would make sure shell extensions are off and then rebooting Windows and trying to uninstall again. Otherwise something has gone horribly bad.

**Comment by Amrsatrio** [comment]
> I do not have Windows Defender for reasons, so none of the files were actually flagged or anything, and I'm not using any other AV program either.

How did you nuke Defender? I have seen reports that an incomplete nuke caused the Defender UI to be inaccessible while the AV itself was still active.

**Comment by LOuroboros** [comment]
> Please go to C:\Program Files\ExplorerPatcher. Are there .prev files? Delete them if there are any.

There aren't.

![%pn_20241122_051308762](https://github.com/user-attachments/assets/1c12ed1c-a2bd-42c3-957e-5c2a889776b1)

> Then check %APPDATA%\ExplorerPatcher and check if there's a folder named cleanup. If there is one, what are the files inside? Try to delete them as well.

No such folder. Only 4 .pdb files; explorer, StartDocked, StartUI and "twinui.pcshell".

> I would make sure shell extensions are off

How do I do that, or what exactly do you mean by that?

> How did you nuke Defender? I have seen reports that an incomplete nuke caused the Defender UI to be inaccessible while the AV itself was still active.

I can't disclose the details, but if it's of any help, the processes that Windows Defender normally use are all inactive. None of them are opened or visible in neither Task Manager nor Process Lasso.

No MsMpEng.exe, no NisSrv.exe, no MpCmdRun.exe, no MpUXSrv.exe and none of the processes in Task Manager mention the word "Defender".

**Comment by Amrsatrio** [comment]
> How do I do that, or what exactly do you mean by that?

Go to EP Properties > File Explorer section. Turn off Register as shell extension and reboot your PC. Then try to uninstall again.

**Comment by Amrsatrio** [comment]
Also, please do me a favor again. Kill explorer.exe, then using Task Manager's "Run New Task" and then "Browse" with file type All Files, go to `C:\Windows\SystemApps\Microsoft.Windows.StartMenuExperienceHost_cw5n1h2txyewy`, and shift+delete these files:

![image](https://github.com/user-attachments/assets/c4da7bc2-a6b6-42fe-b996-eb883d9e7b85)

Also this one in `C:\Windows\SystemApps\ShellExperienceHost_cw5n1h2txyewy`:

![image](https://github.com/user-attachments/assets/9968bfbc-85a9-4eb2-bbd4-653cbd7f69d1)

And lastly this one in `C:\Windows`:

![image](https://github.com/user-attachments/assets/a6474be7-7e0f-44d7-bade-2c2c6449f74f)

**Comment by LOuroboros** [comment]
> > How do I do that, or what exactly do you mean by that?
> 
> Go to EP Properties > File Explorer section. Turn off Register as shell extension and reboot your PC. Then try to uninstall again.

I can't open Explorer Patcher though.

As I mentioned a few times and even showed in a picture, right now my OS is working exactly like stock Windows 11. If I right click my taskbar I just have the Taskbar Settings and the Task Manager.

> Also, please do me a favor again. Kill explorer.exe, then using Task Manager's "Run New Task" and then "Browse" with file type All Files, go to C:\Windows\SystemApps\Microsoft.Windows.StartMenuExperienceHost_cw5n1h2txyewy, screenshot the contents of the folder in your PC to help me find the cause of the problem, and then shift+delete these files:

None of the files you asked me to delete are present, but here's the pic anyway:

![Taskmgr_20241122_081811562](https://github.com/user-attachments/assets/0fa5e89c-103b-42fc-b4fe-38e2e2e42852)

There's no dxgi.dll in neither `C:\Windows\SystemApps\ShellExperienceHost_cw5n1h2txyewy` nor `C:\Windows` either.

**Comment by Amrsatrio** [comment]
Ok, now try shift+deleting C:\Program Files\ExplorerPatcher. Then try uninstalling/unregistering ep_dwm if you installed it. Then go to regedit, delete `HKEY_CURRENT_USER\Software\ExplorerPatcher`. Then delete `HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\{D17F1E1A-5919-4427-8F89-A1A8503CA3EB}_ExplorerPatcher`. This should fully uninstall EP.

I can't recall an issue which may cause the installer to fail.

**Comment by LOuroboros** [comment]
> Ok, now try shift+deleting C:\Program Files\ExplorerPatcher. Then try uninstalling/unregistering ep_dwm if you installed it. Then go to regedit, delete `HKEY_CURRENT_USER\Software\ExplorerPatcher`. Then delete `HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\{D17F1E1A-5919-4427-8F89-A1A8503CA3EB}_ExplorerPatcher`. This should fully uninstall EP.
> 
> I can't recall an issue which may cause the installer to fail.

I didn't have a key called `{D17F1E1A-5919-4427-8F89-A1A8503CA3EB}` at `HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall` but I did the other 2 things.

I guess that's it for the time being. I found a couple of standalone programs to replicate some of the functionality in Explorer Patcher within the past few hours, so I think that right now I'm okay with letting my system breath before trying to install EP again.

Seriously, thanks for all the help!

**Comment by realrelook** [comment]
Hello, I have a similar issue. But I can not delete any of the files listed above e.g. C:\Windows\SystemApps\Microsoft.Windows.StartMenuExperienceHost_cw5n1h2txyewy
even with admin rights.
Any other idea.?
Thank you

**Comment by pyrates999** [comment]
Add the exceptions/exclusions to your antivirus first and then try.

**Comment by realrelook** [comment]
Hello, there is no way out on company level. You should add a warning not to install in Biz environment.
My PC will get new Win11 installation.

**Comment by Amrsatrio** [comment]
@realrelook Do not delete the whole C:\Windows\SystemApps\Microsoft.Windows.StartMenuExperienceHost_cw5n1h2txyewy folder. It is your XAML start menu.

**Comment by pyrates999** [comment]
It's mentioned in the wiki: https://github.com/valinet/ExplorerPatcher/wiki/ExplorerPatcher's-taskbar-implementation#%EF%B8%8F-warning---read-thoroughly-before-using

But perhaps a simple confirmation dialog should be shown when installing it for the first time.

**Comment by Amrsatrio** [comment]
But no one reads the wiki for the first time. Best I could do is to update the release notes generator to always include the message.

**Comment by Amrsatrio** [comment]
@pyrates999 You've been responding to a lot of issues, thanks for that. Can you summarize what is going on with the setup not able to uninstall completely on certain, even fresh machines?

**Comment by pyrates999** [comment]
Here is what I'm noticing:

1. The most common issue is Antivirus blocking files from being copied to the file system.
2. The other one is when there is .prev files, it shows an error message but doesn't tell you where to find the .prev files.
3. When the user upgrades their windows doing a feature upgrade, like windows 11 24h2 to 25h2.  The windows 10 taskbar and windows 10 start menu either doesn't work or doesn't even show in the EP Properties application.  They have to uninstall EP and reinstall it to fix it.  When the EP Properties is opened, it should detect the version of windows you're using and enable the new features, like windows 10 taskbar and windows 10 start menu.  Or if it can't do that, show an error message indicating to uninstall EP and reinstall EP.
4. When antivirus blocks files from being copied over, EP doesn't check that they are there and silently fails.  It should detect they are missing by showing an error message indicating missing files, telling the user to uninstall EP and reinstall it with the latest release.
5. The wiki should have a page for manually removing EP in case the uninstall process doesn't fully work.  I'm not completely sure how to do this, so a page should list the steps needed.  Some users don't even use the EP uninstall method, they use revo uninstaller, which tries to detect additional files and registry entries that should be removed after the initial uninstaller that EP uses.
6. For register as shell, there has to be a warning shown before it's enabled.  It's only used for enabling I think 3 features, but users try to enable it as a way to fix any bug they are encountering, without knowing what it does.
7. The install process should show a confirmation about not installing this on work machines due to it requiring administrative privileges.  It will not be able to update if it can't get administrative privileges and will silently fail I think.  The user should have to confirm this I think.
8. The install should show some kind of progress bar as right now nothing is shown and some users think it is not working.
9. Having a command line option like /q for silent installs would also be helpful.
10. Enabling debug symbol checking after being installed would also be useful as some users can't get the debug symbols the first time it tries.  This can be done manually by running another application or automatically through a service.  It is an improvement right now where it does show you when debug symbols have started downloading and when they finished.
11. A better install method that allows for automatic repairs, like msi would be very useful.
12. When a new release is done, a minimum windows build that is supported would be helpful for those who are trying to install it on an older build of windows.  So that when they install it, an error message is shown instead of it silently failing with it showing crash error messages.
13. When you uninstall it, and you have register as shell enabled, it should detect that and either prevent you from uninstalling so you have to disable it to then uninstall it, or disable that automatically as part of the uninstall process.
14. In several issues, the EP console only shows `started gui thread`, because some antiviruses blocks access to `ddraw.dll`.  Here's the fix for it: https://github.com/valinet/ExplorerPatcher/issues/3377#issuecomment-2176033755.  I think if EP console checks for this showing an error message on how to fix it, that would be more helpful.
15. The weather feature doesn't always work if google detects you're a bot and you can't click past the captcha.  Or if it can't get your weather for where you set the location to.  It also shows the same error message making it look like you're not connected to the internet.
16. Right now the windows 10 battery flyout option doesn't work in windows 11.  It silently crashes.  And even though it is documented in the wiki here: https://github.com/valinet/ExplorerPatcher/wiki/Frequently-asked-questions#the-windows-10-battery-flyout-does-not-open.  It be more useful for the EP properties application to show some warning message when they try to switch to it.  And some do ask why there isn't one for windows 11, which is that windows 11 doesn't have a battery flyout.  But the good news is that it is being worked on here: https://github.com/valinet/ExplorerPatcher/issues/3640

Sorry this is so long, just stuff I've noticed.  I would try to fix some of it, if I was a programmer :)

**Comment by pyrates999** [comment]
Oh and startallback.  It's not compatible with EP, but EP doesn't check if it is installed.  It be useful if it did either before the install, and after if the user decides to try to install startallback with EP already installed.

**Comment by pyrates999** [comment]
Oh and in windows 11, windows 10 taskbar when it is set to the side or the top, after the display resolution changes or the display goes to sleep, windows will appear on top or underneath of it instead of below it, or to the side of it.

**Comment by pyrates999** [comment]
I think the biggest though, is `C:\Windows\dxgi.dll`.  If only `dxgi.dll` could be moved elsewhere in `C:\Program Files\ExplorerPatcher` instead.  But I don't think that can be done due to the way EP works.

**Comment by Amrsatrio** [comment]
Thank you. Your observations are correct. Sorry for my hiatus too, I've been focusing on my thesis amd even freezing updates on my machines so that I'm not distracted with breakages and stuff.

ep_setup.exe really needs to be replaced with an .msi, I should explore options like WiX in the future.

And the dxgi.dll injection method I think is here to stay unless anyone with good knowledge of making injectors is able to make a PR to make a Windhawk-like injector. Wish I can get rid of this myself though.

I'll push the battery flyout fix once I have free time, it's very stable.

**Comment by Amrsatrio** [comment]
Can you list known-bad antiviruses? If the setup runs but the files couldn't be copied with ERROR_ACCESS_DENIED, best I could do is add a warning that certain known-bad antiviruses are present.

**Comment by pyrates999** [comment]
No problem.

SentinelOne is the antivirus that stands out the most.  And when you add an exception to it, the exception only applies if you are a member of the administrators group.  Regular users won't have the exceptions applied to them.  So EP will still appear broken such as windows 10 taskbar not working or windows 10 start menu not working.  I've seen the windows 10 start menu suddenly start working once the exceptions were added to SentinelOne.  So the files were there, they were just being blocked by the antivirus.

I also suggested to include asking what antivirus the user has installed besides windows defender, as part of submitting an issue.  And if they have added the exceptions that EP requires if it's a third party antivirus.

---

Here's a link on how to create a new mod for windhawk that may give you clues on how it works: https://github.com/ramensoftware/windhawk/wiki/Creating-a-new-mod

It does say each mod is a dynamic library.

And someone did create a mod for using the windows 10 taskbar, but only on windows 11 24h2, not 25h2; https://windhawk.net/mods/win10-taskbar-on-win11-24h2

---

And I just remembered.  Someone had submitted a patch that excludes certain exe's from being patched in memory by the Register As Shell feature as it was causing them to crash.  It appears to be from one of the developers for Media Player Classic Home Cinema.  But it's in an issue, not a formal pull request.  Here it is: https://github.com/valinet/ExplorerPatcher/issues/4819

**Comment by Amrsatrio** [comment]
Gotta accept what SentinelOne is doing. Persistence this way is exactly what it's preventing. Idk 🤷

Don't worry, I have two published mods in the WH marketplace. There is an unpublished ep_taskbar_loader mod in the WH server, I could work the original author to make it ready for use by the broader public.

Also regarding that patch, remind me to check before I push the update with the battery flyout fix.
